src/gentropy/dataset/colocalisation.py

```diff
@@ -23,3 +35,105 @@ def get_schema(cls: type[Colocalisation]) -> StructType:
             StructType: Schema for the Colocalisation dataset
         """
         return parse_spark_schema("colocalisation.json")
+
+    def extract_maximum_coloc_probability_per_region_and_gene(
+        self: Colocalisation,
+        study_loci: StudyLocus | L2GGoldStandard,
+        study_index: StudyIndex,
+        filter_by_colocalisation_method: str,
+        filter_by_qtl: str | None = None,
+    ) -> DataFrame:
+        """Get maximum colocalisation probability for a (studyLocus, gene) window.
+
+        Args:
+            study_loci (StudyLocus | L2GGoldStandard): Dataset containing study loci to filter the colocalisation dataset on and the geneId linked to the region
+            study_index (StudyIndex): Study index to use to get study metadata
+            filter_by_colocalisation_method (str): optional filter to apply on the colocalisation dataset
+            filter_by_qtl (str | None): optional filter to apply on the colocalisation dataset
+
+        Returns:
+            DataFrame: table with the maximum colocalisation scores for the provided study loci
+
+        Raises:
+            ValueError: if filter_by_qtl is not in the list of valid QTL types
+            ValueError: if filter_by_colocalisation_method is not in the list of valid colocalisation methods
+        """
+        from gentropy.colocalisation import ColocalisationStep
+
+        valid_qtls = list(EqtlCatalogueStudyIndex.method_to_study_type_mapping.values())
+        if filter_by_qtl and filter_by_qtl not in valid_qtls:
+            raise ValueError(f"There are no studies with QTL type {filter_by_qtl}")
+
+        if filter_by_colocalisation_method not in [
+            "ECaviar",
+            "Coloc",
+        ]:  # TODO: Write helper class to retrieve coloc method names
+            raise ValueError(
+                f"Colocalisation method {filter_by_colocalisation_method} is not supported."
+            )
+
+        method_colocalisation_metric = ColocalisationStep._get_colocalisation_class(
```

Comment:
1. The imports inside the function are a matter to discuss, you usually should place them at the top (as PEP8) suggests it to do. I normally to a think like this when I would introduce a dependency cycle (which should be avoided). But moving this method to the bottom level module seems more appropriate in this case.
2. Having such a method call introduces a risk, when someone tries to import back the `Colocalisation` dataset in the step directly. 

For now it might not be a problem, but it's not standard approach and can cause trouble in the future, that is all :)